File: ab-cloud/verification/matlab/spinor38/spinor38.m
%% spinor38.m — Test 38: 64 spinor structures of the Klein quartic (MATLAB/Octave port)
% Self-implemented cyclic Jacobi eigenvalue algorithm (eig() NOT used).
% Run (Octave): octave spinor38.m [repo-root]   |   MATLAB: spinor38('repo-root')
%
% Validated C++ reference output: isospectrality max|dlambda| = 3.419e-14,
% <r> = 0.4515710793, VERDICT PASS.

function spinor38()
  dd = find_data_dir();
  % ---- classes ----
  fid = fopen(fullfile(dd, 'spinor_classes.csv'), 'r');
  hdr = fgetl(fid);
  classes = {};
  while true
    ln = fgetl(fid);
    if ~ischar(ln), break; end
    if isempty(strtrim(ln)), continue; end
    classes{end+1} = ln; %#ok<AGROW>
  end
  fclose(fid);
  nclasses = numel(classes);
  cls_idx = zeros(nclasses,1); orbit = zeros(nclasses,1); arf = zeros(nclasses,1);
  signs = zeros(nclasses,84);
  for i = 1:nclasses
    parts = strsplit(classes{i}, ',', 'CollapseDelimiters', false);
    cls_idx(i) = str2double(parts{1});
    orbit(i) = str2double(parts{2});
    arf(i) = str2double(parts{3});
    sv = regexp(strtrim(parts{4}), '\s+', 'split');
    for k = 1:84
      signs(i,k) = str2double(sv{k});
    end
  end
  % ---- edges ----
  fid = fopen(fullfile(dd, 'klein_graph_edges.csv'), 'r');
  fgetl(fid);
  edges = [];
  while true
    ln = fgetl(fid);
    if ~ischar(ln), break; end
    if isempty(strtrim(ln)), continue; end
    p = strsplit(strtrim(ln), ',');
    edges = [edges; str2double(p{2}), str2double(p{3})]; %#ok<AGROW>
  end
  fclose(fid);
  % ---- reference stats ----
  fid = fopen(fullfile(dd, 'reference_stats.json'), 'r');
  js = fread(fid, '*char')';
  fclose(fid);
  r_ref = json_num(js, 'r_mean_reference');
  n_zero_ref = round(json_num(js, 'n_zero_modes'));
  representative = round(json_num(js, 'representative_class'));

  N = 56;
  n_odd = sum(orbit == 0);

  % ---- spectra of all odd-orbit classes (self-implemented Jacobi) ----
  k = 0;
  spectra = zeros(N, n_odd);
  rep = zeros(N,1);
  for i = 1:nclasses
    if orbit(i) ~= 0, continue; end
    A = zeros(N, N);
    for e = 1:size(edges,1)
      u = edges(e,1) + 1; v = edges(e,2) + 1;
      A(u,v) = signs(i,e);
      A(v,u) = signs(i,e);
    end
    k = k + 1;
    spectra(:,k) = jacobi_eigen(A);
    if cls_idx(i) == representative
      rep = spectra(:,k);
    end
  end

  % ---- isospectrality ----
  isomax = 0;
  for a = 1:k
    for b = (a+1):k
      isomax = max(isomax, max(abs(spectra(:,a) - spectra(:,b))));
    end
  end

  % ---- zero modes and <r> (fold |lambda|) ----
  lam = abs(rep);
  n_zero = sum(lam < 1e-8);
  lam = sort(lam);
  dsp = diff(lam);
  dsp = dsp(dsp > 1e-8);
  ratios = min(dsp(1:end-1), dsp(2:end)) ./ max(dsp(1:end-1), dsp(2:end));
  r_mean = mean(ratios);

  iok = isomax < 1e-9;
  rok = abs(r_mean - r_ref) < 1e-6;
  ok = iok && rok && (n_zero == n_zero_ref);

  fprintf('Test 38 - 64 spinor structures of the Klein quartic (MATLAB port)\n');
  fprintf('classes loaded: %d | odd-orbit members: %d\n', nclasses, n_odd);
  fprintf('isospectrality within the odd orbit: max|dlambda| = %.3e -> %s\n', ...
          isomax, ternary(iok, 'PASS', 'FAIL'));
  fprintf('zero modes (representative): %d (expected %d)\n', n_zero, n_zero_ref);
  fprintf('<r> (representative): %.10f (reference 0.4515710793) -> %s\n', ...
          r_mean, ternary(rok, 'PASS', 'FAIL'));
  fprintf('VERDICT: %s\n', ternary(ok, 'PASS', 'FAIL'));
  if ~ok, exit(1); end
end

function v = json_num(js, key)
  pat = ['"' key '":'];
  i = strfind(js, pat);
  if isempty(i), v = 0; return; end
  rest = js(i(1)+numel(pat):end);
  tok = strtok(strtrim(rest), ' ,}');
  v = str2double(tok);
end

function out = ternary(cond, a, b)
  if cond, out = a; else, out = b; end
end

function w = jacobi_eigen(Ain)
  N = size(Ain, 1);
  A = Ain;
  for sweep = 1:200
    off = 0;
    for p = 1:N-1
      for q = p+1:N
        off = off + A(p,q)^2;
      end
    end
    if off < 1e-24, break; end
    for p = 1:N-1
      for q = p+1:N
        if abs(A(p,q)) < 1e-15, continue; end
        tau = (A(q,q) - A(p,p)) / (2 * A(p,q));
        t = sign(tau) / (abs(tau) + sqrt(1 + tau^2));
        c = 1 / sqrt(1 + t^2);
        s = t * c;
        for kk = 1:N
          akp = A(kk,p); akq = A(kk,q);
          A(kk,p) = c*akp - s*akq;
          A(kk,q) = s*akp + c*akq;
        end
        for kk = 1:N
          apk = A(p,kk); aqk = A(q,kk);
          A(p,kk) = c*apk - s*aqk;
          A(q,kk) = s*apk + c*aqk;
        end
      end
    end
  end
  w = diag(A)';
  w = sort(w);
end

function dd = find_data_dir()
  dd = pwd();
  for up = 1:6
    cand = fullfile(dd, 'verification', 'spinor64', 'data', 'spinor_classes.csv');
    if exist(cand, 'file'), return; end
    dd = fullfile(dd, '..');
  end
  error('data dir not found; run from inside the repository');
end

File: ab-cloud/verification/julia/spinor38/spinor38.jl
#!/usr/bin/env julia
# spinor38.jl — Test 38: 64 spinor structures of the Klein quartic (Julia port)
# Verifies, using ONLY the frozen data files and the stdlib LinearAlgebra:
#   1. the 28 odd (Arf=1) structures are exactly isospectral (max pairwise
#      spectral distance ~ 1e-14) — no spinor structure is unique;
#   2. <r> of the representative spectrum matches the reference
#      0.4515710792825435.
#
# Run: julia spinor38.jl [repo-root]

using LinearAlgebra
using Printf

function find_data_dir(root)
    bases = String[]
    !isempty(root) && push!(bases, root)
    push!(bases, pwd())
    for b in bases
        for _ in 1:6
            cand = joinpath(b, "verification", "spinor64", "data",
                            "spinor_classes.csv")
            isfile(cand) && return joinpath(b, "verification", "spinor64", "data")
            b = joinpath(b, "..")
        end
    end
    error("data dir not found; pass repo root as argument")
end

function jacobi_eigen(A::Matrix{Float64})
    n = size(A, 1)
    M = copy(A)
    for sweep in 1:200
        off = 0.0
        for p in 1:n, q in (p+1):n
            off += M[p, q]^2
        end
        off < 1e-24 && break
        for p in 1:n, q in (p+1):n
            abs(M[p, q]) < 1e-15 && continue
            tau = (M[q, q] - M[p, p]) / (2 * M[p, q])
            t = sign(tau) / (abs(tau) + sqrt(1 + tau^2))
            c = 1 / sqrt(1 + t^2)
            s = t * c
            for k in 1:n
                akp, akq = M[k, p], M[k, q]
                M[k, p] = c * akp - s * akq
                M[k, q] = s * akp + c * akq
            end
            for k in 1:n
                apk, aqk = M[p, k], M[q, k]
                M[p, k] = c * apk - s * aqk
                M[q, k] = s * apk + c * aqk
            end
        end
    end
    sort!(diag(M))
end

function main()
    root = length(ARGS) >= 1 ? ARGS[1] : ""
    dd = find_data_dir(root)

    classes = Tuple{Int,Int,Int,Vector{Float64}}[]
    lines = readlines(joinpath(dd, "spinor_classes.csv"))
    for ln in lines[2:end]
        isempty(ln) && continue
        parts = split(ln, ','; limit=4)
        cls = parse(Int, parts[1]); orb = parse(Int, parts[2])
        arf = parse(Int, parts[3])
        sgn = [parse(Float64, tok) for tok in split(parts[4])]
        push!(classes, (cls, orb, arf, sgn))
    end
    edges = Tuple{Int,Int}[]
    elines = readlines(joinpath(dd, "klein_graph_edges.csv"))
    for ln in elines[2:end]
        isempty(ln) && continue
        parts = split(ln, ',')
        push!(edges, (parse(Int, parts[2]), parse(Int, parts[3])))
    end
    js = read(joinpath(dd, "reference_stats.json"), String)
    function jsonnum(key::String)
        m = match(Regex("\"$(key)\":\s*([-0-9.eE+]+)"), js)
        isnothing(m) ? 0.0 : parse(Float64, m.captures[1])
    end
    r_ref = jsonnum("r_mean_reference")
    n_zero_ref = round(Int, jsonnum("n_zero_modes"))
    representative = round(Int, jsonnum("representative_class"))

    const_N = 56
    n_odd = count(c -> c[2] == 0, classes)

    spectra = Vector{Vector{Float64}}()
    rep_spectrum = Vector{Float64}()
    for (cls, orb, _arf, sgn) in classes
        orb == 0 || continue
        A = zeros(const_N, const_N)
        for (k, (u, v)) in enumerate(edges)
            A[u+1, v+1] = sgn[k]
            A[v+1, u+1] = sgn[k]
        end
        w = jacobi_eigen(A)
        cls == representative && (rep_spectrum = w)
        push!(spectra, w)
    end

    isomax = 0.0
    for a in 1:length(spectra), b in (a+1):length(spectra), i in 1:const_N
        isomax = max(isomax, abs(spectra[a][i] - spectra[b][i]))
    end

    lam = sort(abs.(rep_spectrum))
    n_zero = count(v -> v < 1e-8, lam)
    dsp = Float64[]
    for i in 1:length(lam)-1
        d = lam[i+1] - lam[i]
        d > 1e-8 && push!(dsp, d)
    end
    r_mean = sum(min(dsp[i], dsp[i+1]) / max(dsp[i], dsp[i+1])
                 for i in 1:length(dsp)-1) / (length(dsp) - 1)

    println("Test 38 - 64 spinor structures of the Klein quartic (Julia port)")
    @printf("classes loaded: %d | odd-orbit members: %d\n", length(classes), n_odd)
    println("isospectrality within the odd orbit: max|dlambda| = ",
            @sprintf("%.3e", isomax), " -> ", isomax < 1e-9 ? "PASS" : "FAIL")
    println("zero modes (representative): ", n_zero, " (expected ", n_zero_ref, ")")
    rok = abs(r_mean - r_ref) < 1e-6
    println("<r> (representative): ", @sprintf("%.10f", r_mean),
            " (reference 0.4515710793) -> ", rok ? "PASS" : "FAIL")
    ok = isomax < 1e-9 && n_zero == n_zero_ref && rok
    println("VERDICT: ", ok ? "PASS" : "FAIL")
    ok || exit(1)
end

main()
